src/NHibernate/Logging.cs

```diff
@@ -316,118 +738,254 @@ static Log4NetLogger()
+		}
+
+		public static void Info(this INHibernateLogger logger, Exception exception, string format, params object[] args)
+		{
+			logger.Log(InternalLogLevel.Info, new InternalLogValues(format, args), exception);
+		}
+
+		public static void Info(this INHibernateLogger logger, string format, params object[] args)
+		{
+			logger.Log(InternalLogLevel.Info, new InternalLogValues(format, args), null);
+		}
+
+		public static void Debug(this INHibernateLogger logger, Exception exception, string format, params object[] args)
+		{
+			logger.Log(InternalLogLevel.Debug, new InternalLogValues(format, args), exception);
+		}
+
+		public static void Debug(this INHibernateLogger logger, string format, params object[] args)
+		{
+			logger.Log(InternalLogLevel.Debug, new InternalLogValues(format, args), null);
+		}
+
+
+		// catch any method calls with an Exception argument second as they would otherwise silently be consumed by `params object[] args`.
+
+		/// <summary>
+		/// Throws NotImplementedException. Calling this method is an error. Please use methods taking the exception as first argument instead.
+		/// </summary>
+		[Obsolete("Use Fatal(Exception, string, params object[])", true)]
+		public static void Fatal(this INHibernateLogger logger, string message, Exception ex) => ThrowNotImplemented();
+
+		/// <summary>
+		/// Throws NotImplementedException. Calling this method is an error. Please use methods taking the exception as first argument instead.
+		/// </summary>
+		[Obsolete("Use Error(Exception, string, params object[])", true)]
+		public static void Error(this INHibernateLogger logger, string message, Exception ex) => ThrowNotImplemented();
+
+		/// <summary>
+		/// Throws NotImplementedException. Calling this method is an error. Please use methods taking the exception as first argument instead.
+		/// </summary>
+		[Obsolete("Use Warn(Exception, string, params object[])", true)]
+		public static void Warn(this INHibernateLogger logger, string message, Exception ex) => ThrowNotImplemented();
+
+		/// <summary>
+		/// Throws NotImplementedException. Calling this method is an error. Please use methods taking the exception as first argument instead.
+		/// </summary>
+		[Obsolete("Use Info(Exception, string, params object[])", true)]
+		public static void Info(this INHibernateLogger logger, string message, Exception ex) => ThrowNotImplemented();
+
+		/// <summary>
+		/// Throws NotImplementedException. Calling this method is an error. Please use methods taking the exception as first argument instead.
+		/// </summary>
+		[Obsolete("Use Debug(Exception, string, params object[])", true)]
+		public static void Debug(this INHibernateLogger logger, string message, Exception ex) => ThrowNotImplemented();
+
+		private static void ThrowNotImplemented()
+		{
+			throw new NotImplementedException("Should not have compiled with call to this method");
+		}
+	}
+
+	public struct InternalLogValues
```

Comment:
This is almost identical tothe ` FormattableString`

Reply:
> This is almost identical to the `FormattableString`

Do you mean `InternalLogValues` should be a class deriving from `FormattableString` instead? Why not. Looking at the concrete implementation available in .Net Framework, it would require slightly more code though.

From `System.Runtime.CompilerServices`:
```c#
private sealed class ConcreteFormattableString : FormattableString
{
	private readonly string _format;

	private readonly object[] _arguments;

	public override string Format
	{
		get
		{
			return _format;
		}
	}

	public override int ArgumentCount
	{
		get
		{
			return _arguments.Length;
		}
	}

	internal ConcreteFormattableString(string format, object[] arguments)
	{
		_format = format;
		_arguments = arguments;
	}

	public override object[] GetArguments()
	{
		return _arguments;
	}

	public override object GetArgument(int index)
	{
		return _arguments[index];
	}

	public override string ToString(IFormatProvider formatProvider)
	{
		return string.Format(formatProvider, _format, _arguments);
	}
}
```